sqlglot/parser.py

```diff
@@ -3156,7 +3156,9 @@ def _parse_id_var(
                 prefix += self._prev.text
 
         if (any_token and self._advance_any()) or self._match_set(tokens or self.ID_VAR_TOKENS):
-            return exp.Identifier(this=prefix + self._prev.text, quoted=False)
+            quoted = self._prev.token_type == TokenType.STRING
+            return exp.Identifier(this=prefix + self._prev.text, quoted=quoted)
```

Comment:
maybe it should instead be

exp.identifier(...)

Reply:
I think we'll also need to do `token_type == TokenType.STRING` for `exp.to_identifier`. If we just do

```
return exp.to_identifier(prefix + self._prev.text)
```

, there are several tests that break:

```
AssertionError: 'X((y AS z))."1"' != 'X((y AS z)).1'
- X((y AS z))."1"
?             - -
+ X((y AS z)).1
```

```
AssertionError: 'CREATE PROCEDURE foo @a INTEGER, @b INTEGER AS SELECT @a = SUM[17 chars] bar' != 'CREATE PROCEDURE foo "@a" INTEGER, "@b" INTEGER AS SELECT "@a"[23 chars] bar'
- CREATE PROCEDURE foo @a INTEGER, @b INTEGER AS SELECT @a = SUM(bla) FROM baz AS bar
+ CREATE PROCEDURE foo "@a" INTEGER, "@b" INTEGER AS SELECT "@a" = SUM(bla) FROM baz AS bar
?                      +  +          +  +                   +  +
```

The tests that fail contain identifiers with special characters like @, $ and numbers in the front, because these are not captured by `SAFE_IDENTIFIER_RE` and so are force-quoted.

Reply:
ok forget it then